README.org
#+TITLE: syncthing-activity

This is a fork of [[https://github.com/jpmens/syncthing-activity][syncthing-activity]] by jpmens. The main difference between this
fork and the original is the support parse relevant information to a script.

This small program uses [[https://syncthing.net][Syncthing]]'s [[https://docs.syncthing.net/dev/rest.html][REST API]] to determine the changes to files
and directories occurring on the local instance of Syncthing. The relevant event
API are [[https://docs.syncthing.net/events/itemfinished.html][ItemFinished]] and [[https://docs.syncthing.net/events/itemstarted.html][ItemStarted]]. It also allows users to connect the API
results to external scripts. 

** Table of contents :TOC_3_gh:
  - [[#setup][Setup]]
    - [[#requirements][Requirements]]
    - [[#installation][Installation]]
    - [[#api-key][API key]]
  - [[#usage][Usage]]
    - [[#events][Events]]
    - [[#environments][Environments]]
    - [[#scripts][Scripts]]
  - [[#example][Example]]
  - [[#docker][docker]]

** Setup
=syncthing-activity= is not hard to run for relatively modern machines. However,
if you are planning to run it on a headless server, I highly recommend running
it as a Docker container, instructions [[#docker][below]].

You will need to have knowledge of your Syncthing instance's URL as well as its
[[#api-key][API key]].


*** Requirements
- Python 3.6 or later

*** Installation
With a working Syncthing instance, simply clone this repository or download the
script file as-is.
#+BEGIN_EXAMPLE shell
  git clone https://github.com/arifer612/syncthing-activity
#+END_EXAMPLE

Install the necessary dependencies using =pip= while in the directory.
#+BEGIN_EXAMPLE shell
  python3 -m pip install -r requirements.txt
#+END_EXAMPLE

*** API key
Open Syncthing's Web UI at =http://127.0.0.1:8384=, click on /Actions/
and /Settings/. On the Settings panel, /General/ tab you'll find the API
key on the right. If your Syncthing is hosted on a different machine or port,
change the values accordingly.

** Usage
The help dialog explains everything you need to know.
#+BEGIN_SRC shell :results output :exports results
  ./syncthing_activity.py -h
#+END_SRC

#+RESULTS:
#+BEGIN_EXAMPLE
usage: syncthing_activity.py [-h] [--event [EVENT]] [--url URL] [--api API]
                             [--script [SCRIPT]]

syncthing-activity v1.1.2 --- Watch a Syncthing server.

optional arguments:
  -h, --help         show this help message and exit
  --event [EVENT]    Event API type. `ItemStarted' or `ItemFinished'.

Environments:
  Over-ride global environment variables.

  --url URL          Syncthing URL, defaults to http://localhost:8384. Using
                     the SYNCTHING_URL environment.
  --api API          Syncthing API key. Using the SYNCTHING_API environment.

Scripts:
  Run a script along the watcher.

  --script [SCRIPT]  Path to the script along with all the required arguments.
                     For example: post-processor.py positional_arg1 --option_1
                     option_arg_1
#+END_EXAMPLE

*** Events
=syncthing-activity=, when executed as-is, behaves just as the original
repository except with the event API =ItemFinished= instead of =ItemStarted=. It
can be reverted by specifying =ItemStarted= using the =--event= flag.
#+BEGIN_EXAMPLE shell
  ./syncthing-activity.py --event "ItemStarted"
#+END_EXAMPLE
The event APIs that can be used as positional arguments are =ItemFinished= and =ItemStarted=.

*** Environments
As with the [[https://github.com/jpmens/syncthing-activity#apikey][original repository]], you can set the Syncthing API key and URL as
environment variables before executing =syncthing-activity=.

#+BEGIN_EXAMPLE
  export SYNCTHING_API="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
  export SYNCTHING_URL="http://localhost:8384"
  ./syncthing-activity.py
#+END_EXAMPLE

If you do not wish to directly export it as an environment variable or for
Windows users, there are environment flags that can be used instead.
#+BEGIN_EXAMPLE
  python3 -m syncthing-activity --url https://localhost:8384 --api xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
#+END_EXAMPLE

*** Scripts
If the purpose is to parse the API results to a script for processing, you can
do so by making use of the =--script= option. Additional arguments are passed
through as-is after specifiying the script.

For example, the following snippet will use the =ItemFinished= event API and
pass the positional argument =positional_arg1=, option =--option1= as well as
its argument =option1_arg=, and the flag =--flag1= to the script at path
=~/scripts/syncthing-postprocess=.
#+BEGIN_EXAMPLE shell
  syncthing-activity.py --script "~/scripts/syncthing-postprocess" positonal_arg1 --option1 option1_arg --flag1
#+END_EXAMPLE

To get your script to work with =syncthing-activity=, it must be made to accept
an argument with the option =--payload=. The contents will be a JSON object
defined as such:
#+BEGIN_EXAMPLE json
  {
      "time": Event time in ISO format YYYY-MM-DDThh:mm:ss.sTZD.
      "action": Event action (update, delete, metadata).
      "type": Data type (file, dir).
      "item": Relative path to item from folder root.
      "error": Sync error result (null, error message).
      "folder_label": Human readable folder label.
      "folder_id": Machine readable folder id.
      "path": Absolute path to item.
  }
#+END_EXAMPLE

External Python scripts with dependencies should come with a =requirements.txt=
that gets mounted along with the script into the =/scripts= directory of the
container. It may also be a good idea to include a dry-run flag in your script
before putting it into production.

** Example
In the absence of a script, the program currently outputs the folder label in
which an update is detected, the object (file or directory) and type of update
as well as the object's name. Passing a script, it can be reflected in the
output as well.

#+BEGIN_EXAMPLE
  owntracks file   update     platform/ansible/templates/config.f
            script mover.py   Moving item config.f to directory at ~/archives
       take file   update     configs/contacts/mac/.git/index
            script mover.py   Moving item index to directory at ~/archives
       take file   metadata   configs/contacts/mac/ab.json
       take file   update     playground/syncthing/events/requirements.txt
            script mover.py   Moving item requirements.txt to directory at ~/archives
       take file   update     playground/syncthing/events/syncthing-activity.py
            script mover.py   Moving item syncthing-activity.py to directory at ~/archives
  on-github file   update     owntracks/recorder/Changelog
            script mover.py   Moving item Changelog to directory at ~/archives
       take dir    update     playground/syncthing/events/docs
            script mover.py   Creating directory docs at ~/archives
       take file   delete     playground/syncthing/events/menu
            script mover.py   Moving item menua to directory at ~/archives
#+END_EXAMPLE

** docker
This mini-application can be deployed through a Docker image. You can retrieve
the image from Dockerhub using
#+BEGIN_EXAMPLE shell
  docker pull arifer612/syncthing-activity:latest
#+END_EXAMPLE

When running with =docker run=, feed the necessary environment variables using
the =-e= flag. Environments:
- =EVENT_API= : Syncthing event API. Currently supports only =ItemFinished=
  (default) and =ItemStarted=
- =SYNCTHING_URL= : URL to the Syncthing instance. The default is
  =http://localhost:8384=
- =SYNCTHING_API= : API key to the Syncthing instance.
- =SCRIPT=: Name of the script after it has been mounted into the container in
  the ~/script~ directory.
- =ARGUMENTS=: Additional arguments that can be passed into the script.

Example:
#+BEGIN_EXAMPLE shell
  docker run -d \
         -v ~/scripts/:/scripts:rw \
         -e "API_TYPE=ItemFinished"
         -e "SYNCTHING_URL=http://localhost:8384" \
         -e "SYNCTHING_API=xxxxx" \
         -e "SCRIPT=syncthing-postprocess" \
         -e "ARGUMENTS=positional_arg1 --option1 option1_arg1 --flag1" \
         arifer612/syncthing-activity:latest
#+END_EXAMPLE


CHANGELOG.org
#+TITLE: CHANGELOG

** v1.1.2 [2022-04-19]

*** Additions
- flake8 ignore file to suppress line-too-long warnings while linting.
- GitHub Actions for linting and documentation building. The CHANGELOG and
  VERSION files are also cleaned up.

*** Changes
- Phase out print in favour for logging of the Python script.
- Python script status metadata is determined dynamically depending on the
  version. If it is hyphenated, e.g.'-rcX' or '-betaY' etc., then it should not
  be on the master branch and is in development. Otherwise, it is in production.

** v1.1.1 [2022-04-17]

*** Changes
- CHANGELOG order and looks, use Org inactive timestamps

*** Fixes
- Handle connection error when Syncthing is restarted
- Handle timeout error when Syncthing is first started

** v1.1.0 [2022-04-14] 

*** Additions
- Git and Docker ignore files for Python cache files
- VERSION file for updating Python script and Docker image versions
- CHANGELOG file

*** Changes
- Python script formatted according to PEP 8 and PEP 257
- Python script made cross-platform by removing direct need for environment variables
- README file updated with example for scripts
- PYTHONUNBUFFERED environment to ensure output of print commands
- Refrain from tracking from the start of time
- Only build and publish Docker images on tags
** v1.0.1 [2022-04-09]

*** Fixes
- Remove hardcoded boolean values

** v1.0.0 [2022-04-09]

*** Additions
- Dockerfile
- GitHub workflow for the building and publishing of Dockerfile

*** Changes
- Compatibility with external scripts

** fork [2022-04-08]
- Forked from parent repository at [[https://github.com/jpmens/syncthing-activity]]. 


syncthing-activity.py
#!/usr/bin/env python3
"""Watches a Syncthing instance's activity."""

import argparse
import json
import os
import subprocess
import sys
import time
import warnings
import socket

import requests

__author__ = "Jan-Piet Mens <jp@mens.de>"
__copyright__ = "Copyright 2019 Jan-Piet Mens"
__license__ = "GNU General Public License"
__maintainer__ = "Arif Er <arifer612@pm.me>"
__status__ = "Development"
with open(
        os.path.join(os.path.dirname(os.path.realpath(__file__)), "VERSION"),
        "r", encoding="UTF-8"
) as version:
    __version__ = version.read().strip()

# Global variables
LAST_ID = 0
FOLDERS = {}

# Environment section
_SYNCTHING_URL = os.getenv("SYNCTHING_URL", "http://localhost:8384")
_SYNCTHING_API = os.getenv("SYNCTHING_API")
_HEADERS = {"X-API-Key": _SYNCTHING_API}


# argparse section
parser = argparse.ArgumentParser(
    description=(f"""
    syncthing-activity {__version__}  ---  Watch a Syncthing server.
    """),
)
parser.add_argument(
    "--event",
    type=str,
    nargs="?",
    const="ItemFinished",
    default="ItemFinished",
    help="Event API type. `ItemStarted' or `ItemFinished'.",
)
# Environment arguments
envs = parser.add_argument_group(
    "Environments", "Over-ride global environment variables."
)
envs.add_argument(
    "--url",
    type=str,
    default=f"{_SYNCTHING_URL}",
    help="Syncthing URL, defaults to http://localhost:8384. \
    Using the SYNCTHING_URL environemnt.",
)
envs.add_argument(
    "--api",
    type=str,
    default=f"{_SYNCTHING_API}",
    help="Syncthing API key. Using the SYNCTHING_API environment.",
)

# Script arguments
scripts = parser.add_argument_group(
    "Scripts", "Run a script along the watcher."
)
scripts.add_argument(
    "--script",
    type=str,
    nargs="?",
    const="",
    default="",
    help="Path to the script along with all the required arguments. \
    For example: post-processor.py positional_arg1 --option_1 option_arg_1"
)

ARGS = argparse.Namespace()
UNKNOWN_ARGS = []


def getfolders():
    """Retrieves folder labels and paths from Syncthing."""
    global FOLDERS

    FOLDERS = {}

    response = requests.get(f"{ARGS.url}/rest/system/config",
                            headers=_HEADERS)
    data = json.loads(response.text)
    for folder in data["folders"]:
        FOLDERS[folder["id"]] = {
            "label": folder["label"],
            "path": folder["path"],
        }


def activeSyncthing() -> bool:
    """Checks if Syncthing is active."""
    protocol, address = ARGS.url.split("://")
    if ":" in address:
        ip, port = address.split(":")
    else:
        ip = address
        try:
            port = socket.getservbyname(protocol)
        except OSError:
            port = 80

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        sock.connect((ip, int(port)))
        sock.shutdown(2)
        return True
    except Exception:
        return False


def process(array: dict, args: argparse.Namespace, unknown_args: list) -> None:
    """Parses API results into usable information.

    With a script, a payload will be passed as a JSON dictionary. The structure
    of the payload is:
        {
      "time": Event time in ISO format YYYY-MM-DDThh:mm:ss.sTZD.
      "action": Event action (update, delete, metadata).
      "type": Data type (file, dir).
      "item": Relative path to item from folder root.
      "error": Sync error result (null, error message).
      "folder_label": Human readable folder label.
      "folder_id": Machine readable folder id.
      "path": Absolute path to item.
    }

    Args:
        array: Syncthing API response.
        args: This script's keyword and positional arguments.
        unknown_args: The remaining arguments not declared in parser and are to
    be passed into the external script.

    Raises:
        ResourceWarning: If a new folder is created on Syncthing and cannot be
            updated here. Restarting the Syncthing instance may resolve this.
    """
    global LAST_ID

    for event in array:
        if "type"in event and event["type"] == args.event:
            LAST_ID = event["id"]

            folder_id = event["data"]["folder"]
            if folder_id not in FOLDERS:  # Re-cache folder list
                getfolders()

            try:
                folder_label = FOLDERS[folder_id]["label"]
                folder_path = FOLDERS[folder_id]["path"]

                payload = {
                    "time": event["time"],
                    "action": event["data"]["action"],
                    "type": event["data"]["type"],
                    "item": event["data"]["item"],
                    "error": event["data"]["error"],
                    "folder_label": folder_label,
                    "folder_id": folder_id,
                    "path": os.path.join(folder_path, event["data"]["item"]),
                }

                print(
                    "{folder_label:>15} {type:<6s} {action:<10s} "
                    "{item}".format(**payload)
                )

                if args.script and args.script != "None":
                    # Note that syncs with errors will still make it here and
                    # need to be handled in external scripts by looking at the
                    # 'error' keyword of the payload.
                    main_call = [
                        sys.executable,
                        args.script,
                        "--payload",
                        json.dumps(payload, indent=4),
                    ]
                    if unknown_args:
                        main_call += unknown_args

                    subprocess.call(main_call)
            except KeyError:
                warnings.warn(f"WARNING! Folder ID {folder_id} cannot be "
                              "accessed. Skipping this.", ResourceWarning)


def main(args: tuple = None) -> None:
    """Main execution block.

    Args:
        args: Tuple of known arguments declared in parser and the remaining
    unknown arguments.
    """
    global ARGS, UNKNOWN_ARGS, _HEADERS, LAST_ID

    print("... Starting up syncthing-activity ...")

    ARGS, UNKNOWN_ARGS = args or parser.parse_known_args()

    if ARGS.api is None:
        print("Missing SYNCTHING_API in environment", file=sys.stderr)
        sys.exit(2)

    print(f"... Connecting to Syncthing hosted at {ARGS.url} ...")

    if not activeSyncthing():
        print("... Syncthing is not running. Stopping the watcher. ...")
        sys.exit(0)

    _HEADERS = {"X-API-Key": ARGS.api}
    params = {
        "since": LAST_ID,
        "limit": None,
        "events": ARGS.event,
    }

    # Retrieve event ID when syncthing-activity starts up instead of starting
    # from 0 if possible.

    # Set the timeout to 5s to account for situations when Syncthing is just
    # started and do not have any recorded events matching ARGS.event.
    try:
        first_response = requests.get(
            f"{ARGS.url}/rest/events", headers=_HEADERS, params=params,
            timeout=5
        )
        if first_response.status_code == 200:
            LAST_ID = json.loads(first_response.content)[-1].get("id")
    except requests.exceptions.ConnectionError:
        pass
    except Exception as err:
        print("... Unknown first connection error: ", err, " ...")
        print("... Continuing with connection ...")

    print("... Successfully connected to Syncthing ...")
    print(f"... {LAST_ID} {ARGS.event} events occurred in the past ...")
    print("... Starting watcher process ...")

    getfolders()

    while True:
        params = {
            "since": LAST_ID,
            "limit": None,
            "events": ARGS.event,
        }

        try:
            response = requests.get(
                f"{ARGS.url}/rest/events", headers=_HEADERS, params=params
            )
            if response.status_code == 200:
                process(json.loads(response.text), ARGS, UNKNOWN_ARGS)
            elif response.status_code != 304:
                time.sleep(60)
                continue
            time.sleep(10.0)
        except requests.exceptions.ChunkedEncodingError:
            # Check if Syncthing is restarted.
            time.sleep(10.0)
            if not activeSyncthing():
                print("... Syncthing may not be running. Stopping the watcher. ...")
                sys.exit(1)


if __name__ == "__main__":
    try:
        main(parser.parse_known_args())
    except KeyboardInterrupt:
        sys.exit(0)


syncthing_activity.py
#!/usr/bin/env python3
"""Watches a Syncthing instance's activity."""

import argparse
import json
import logging
import os
import socket
import subprocess
import sys
import time
from types import TracebackType

import requests

__author__ = "Jan-Piet Mens <jp@mens.de>"
__copyright__ = "Copyright 2019 Jan-Piet Mens"
__license__ = "GNU General Public License"
__maintainer__ = "Arif Er <arifer612@pm.me>"
with open(
    os.path.join(os.path.dirname(os.path.realpath(__file__)), "VERSION"),
    "r",
    encoding="UTF-8",
) as version:
    __version__ = version.read().strip()
    if "-" in __version__:
        __status__ = "Development"
    else:
        __status__ = "Production"


# Global variables
LAST_ID = 0
FOLDERS = {}
ARGS = argparse.Namespace()
UNKNOWN_ARGS = []


# Environment section
_SYNCTHING_URL = os.getenv("SYNCTHING_URL", "http://localhost:8384")
_SYNCTHING_API = os.getenv("SYNCTHING_API", "")
_HEADERS = {"X-API-Key": _SYNCTHING_API}


# argparse section
parser = argparse.ArgumentParser(
    description=(f"syncthing-activity {__version__}  ---  Watch a Syncthing server."),
)
parser.add_argument(
    "--event",
    type=str,
    nargs="?",
    const="ItemFinished",
    default="ItemFinished",
    help="Event API type. `ItemStarted' or `ItemFinished'.",
)
# Environment arguments
envs = parser.add_argument_group(
    "Environments", "Over-ride global environment variables."
)
envs.add_argument(
    "--url",
    type=str,
    default=f"{_SYNCTHING_URL}",
    help="Syncthing URL, defaults to http://localhost:8384. \
    Using the SYNCTHING_URL environment.",
)
envs.add_argument(
    "--api",
    type=str,
    default=f"{_SYNCTHING_API}",
    help="Syncthing API key. Using the SYNCTHING_API environment.",
)
# Script arguments
scripts = parser.add_argument_group("Scripts", "Run a script along the watcher.")
scripts.add_argument(
    "--script",
    type=str,
    nargs="?",
    const="",
    default="",
    help="Path to the script along with all the required arguments. \
    For example: post-processor.py positional_arg1 --option_1 option_arg_1",
)


# Logger
logger = logging.getLogger(__name__)
handler = logging.StreamHandler(stream=sys.stdout)
formatter = logging.Formatter(
    "%(asctime)s :: %(levelname)-8s :: %(message)s", datefmt="%Y-%m-%d %H:%M:%S"
)
handler.setFormatter(formatter)
logger.setLevel(logging.INFO)
logger.addHandler(handler)


def handle_exception(
    exc_type: type, exc_value: int, exc_traceback: TracebackType
) -> None:
    """Log any unexpected exceptions through the logger.

    Retrieved from https://stackoverflow.com/a/16993115
    """
    if issubclass(exc_type, KeyboardInterrupt):
        sys.__excepthook__(exc_type, exc_value, exc_traceback)
        return
    logger.exception(
        "Unexpected exception", exc_info=(exc_type, exc_value, exc_traceback)
    )


def get_folders() -> None:
    """Retrieves folder labels and paths from Syncthing."""
    global FOLDERS

    FOLDERS = {}

    response = requests.get(f"{ARGS.url}/rest/system/config", headers=_HEADERS)
    data = json.loads(response.text)
    for folder in data["folders"]:
        FOLDERS[folder["id"]] = {
            "label": folder["label"],
            "path": folder["path"],
        }


def active_syncthing() -> bool:
    """Checks if Syncthing is active."""
    protocol, address = ARGS.url.split("://")
    if not address:
        protocol, address = "http", protocol

    if ":" in address:
        host, port = address.split(":")
    else:
        host = address
        try:
            port = socket.getservbyname(protocol)
        except OSError:
            port = 80

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        sock.connect((host, int(port)))
        sock.shutdown(2)
        return True
    except (ConnectionRefusedError, OSError):
        return False


def close_logger(exit_code: int) -> None:
    """Graciously closes the logger and ends the watcher.

    Args:
        exit_code: System exit code
    """
    logger.info("Ending syncthing-activity")
    logging.shutdown()
    sys.exit(exit_code)


def process(array: dict, args: argparse.Namespace, unknown_args: list) -> None:
    """Parses API results into usable information.

    With a script, a payload will be passed as a JSON dictionary. The structure
    of the payload is:
        {
      "time": Event time in ISO format YYYY-MM-DDThh:mm:ss.sTZD.
      "action": Event action (update, delete, metadata).
      "type": Data type (file, dir).
      "item": Relative path to item from folder root.
      "error": Sync error result (null, error message).
      "folder_label": Human readable folder label.
      "folder_id": Machine readable folder id.
      "path": Absolute path to item.
    }

    Args:
        array: Syncthing API response.
        args: This script's keyword and positional arguments.
        unknown_args: The remaining arguments not declared in parser and are to
    be passed into the external script.
    """
    global LAST_ID

    for event in array:
        if "type" in event and event["type"] == args.event:
            LAST_ID = event["id"]

            folder_id = event["data"]["folder"]
            if folder_id not in FOLDERS:  # Re-cache folder list
                get_folders()

            try:
                folder_label = FOLDERS[folder_id]["label"]
                folder_path = FOLDERS[folder_id]["path"]

                payload = {
                    "time": event["time"],
                    "action": event["data"]["action"],
                    "type": event["data"]["type"],
                    "item": event["data"]["item"],
                    "error": event["data"]["error"],
                    "folder_label": folder_label,
                    "folder_id": folder_id,
                    "path": os.path.join(folder_path, event["data"]["item"]),
                }

                logger.info(
                    "%(folder_label)15s %(type)-6s %(action)-10s %(item)s", payload
                )

                if args.script and args.script != "None":
                    # Note that syncs with errors will still make it here and
                    # need to be handled in external scripts by looking at the
                    # 'error' keyword of the payload.
                    main_call = [
                        sys.executable,
                        args.script,
                        "--payload",
                        json.dumps(payload, indent=4),
                    ]
                    if unknown_args:
                        main_call += unknown_args

                    subprocess.call(main_call)
            except KeyError:
                logger.warning(
                    "WARNING! Folder ID %s cannot be accessed. Skipping this.",
                    folder_id,
                )


def main(args: tuple = None) -> None:
    """Main execution block.

    Args:
        args: Tuple of known arguments declared in parser and the remaining
    unknown arguments.
    """
    global ARGS, UNKNOWN_ARGS, _HEADERS, LAST_ID

    sys.excepthook = handle_exception
    logger.info("Starting up syncthing-activity")

    ARGS, UNKNOWN_ARGS = args or parser.parse_known_args()

    if not ARGS.api:
        logger.error("Missing SYNCTHING_API in environment")
        close_logger(2)

    logger.info("Connecting to Syncthing hosted at %s", ARGS.url)

    if not active_syncthing():
        logger.info("Syncthing is not running. Stopping the watcher.")
        close_logger(127)

    _HEADERS = {"X-API-Key": ARGS.api}
    params = {
        "since": LAST_ID,
        "limit": None,
        "events": ARGS.event,
    }

    # Retrieve event ID when syncthing-activity starts up instead of starting
    # from 0 if possible.

    # Set the timeout to 5s to account for situations when Syncthing is just
    # started and do not have any recorded events matching ARGS.event.
    try:
        first_response = requests.get(
            f"{ARGS.url}/rest/events", headers=_HEADERS, params=params, timeout=5
        )
        if first_response.status_code == 200:
            LAST_ID = json.loads(first_response.content)[-1].get("id")
    except requests.exceptions.ConnectionError:
        # If timeout, then LAST_ID is 0 and needs not be changed.
        pass

    logger.info("Successfully connected to Syncthing")
    logger.info("%d %s events occurred in the past", LAST_ID, ARGS.event)
    logger.info("Starting watcher process")

    get_folders()

    while True:
        params = {
            "since": LAST_ID,
            "limit": None,
            "events": ARGS.event,
        }

        try:
            response = requests.get(
                f"{ARGS.url}/rest/events", headers=_HEADERS, params=params
            )

            if response.status_code == 200:
                process(json.loads(response.text), ARGS, UNKNOWN_ARGS)
            elif response.status_code != 304:
                time.sleep(60)
                continue
            time.sleep(10.0)

        except requests.exceptions.ChunkedEncodingError:
            # Check if Syncthing is restarted.
            time.sleep(10.0)
            if not active_syncthing():
                logger.info("Syncthing may not be running. Stopping the watcher.")
                close_logger(127)


if __name__ == "__main__":
    try:
        main(parser.parse_known_args())
    except KeyboardInterrupt:
        sys.stderr.write("\r")  # Clear ^C from showing
        close_logger(0)
